File: 78. Subsets/recursion2.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        LinkedList<List<Integer>> res = new LinkedList<List<Integer>>();
        if (nums == null || nums.length == 0) return res;
        helper(nums, 0, res, new LinkedList<Integer>());
        return res;
    }
    public void helper(int[] nums, int curidx, List<List<Integer>> res, List<Integer> cur){ 
        if (curidx == nums.length){
            res.add(new LinkedList<Integer>(cur));
        }
        else{
            //保留原来的
            helper(nums, curidx + 1, res, cur);
            //加上下一个
            cur.add(nums[curidx]);
            helper(nums, curidx + 1, res, cur);
            cur.remove(cur.size() - 1);
        }
    }
}

File: 78. Subsets/recursion2.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def recursion(r, ci, cl): # current result, current index, current list
            if ci == len(nums):
                r.append(cl[:])
            else:
                recursion(r, ci+1, cl)
                cl.append(nums[ci])
                recursion(r, ci+1, cl)   # 根本不需要用for循环，因为这里调用的时候是加1调用，这也是一个递归不需要用到循环的很好的例子，因为本身就是循环的思想
                cl.pop()
        
        r, cl = [], []
        recursion(r, 0, cl)
        return r